File: PipeServerTest/pipe_comm.cpp
#include <iostream>
#include <Windows.h>
#include <stdio.h>
#include <thread>
#include <vector>
#include <string>


#define BUFFERSIZE 1024
#define ADDIN_STREAM L"\\\\.\\pipe\\ADDIN_DATA_STREAM"
#define TENSOR_STREAM L"\\\\.\\pipe\\TENSORFLOW_DATA_STREAM"

void Read();
void Write();

HANDLE hPipeR, hPipeW;


bool quit = false;

int main() {
	std::vector<std::thread> threads;

	threads.push_back(std::thread(Read));
	threads.push_back(std::thread(Write));

	for (auto& th : threads) th.join();

	return 0;
}


void Read() {
	hPipeR = INVALID_HANDLE_VALUE;
	char buffer[1024];
	DWORD dwRead;

	hPipeR = CreateNamedPipe(TENSOR_STREAM,
		PIPE_ACCESS_DUPLEX,
		PIPE_TYPE_BYTE | PIPE_READMODE_BYTE | PIPE_WAIT,   // FILE_FLAG_FIRST_PIPE_INSTANCE is not needed but forces CreateNamedPipe(..) to fail if the pipe already exists...
		1,
		BUFFERSIZE,
		BUFFERSIZE,
		NMPWAIT_USE_DEFAULT_WAIT,
		NULL);

	while (hPipeR != INVALID_HANDLE_VALUE && !quit)
	{
		if (ConnectNamedPipe(hPipeR, NULL) != FALSE)   // wait for someone to connect to the pipe
		{
			while (ReadFile(hPipeR, buffer, sizeof(buffer) - 1, &dwRead, NULL) != FALSE)
			{
				/* add terminating zero */
				buffer[dwRead] = '\0';

				/* do something with data in buffer */
				printf("DATA RECV: %s", buffer);

			}
		}

		DisconnectNamedPipe(hPipeR);
	}
	CloseHandle(hPipeR);
	printf("CPP - Read Stream Closed\n");
}

void Write() {
	hPipeW = INVALID_HANDLE_VALUE;
	DWORD dwWritten;

	//Attempt to connect to pipe
	while (hPipeW == INVALID_HANDLE_VALUE) {
		hPipeW = CreateFile(ADDIN_STREAM,
			GENERIC_READ | GENERIC_WRITE,
			0,
			NULL,
			OPEN_EXISTING,
			0,
			NULL);
	}

	while (!quit) {
		printf("Write Message\n");
		std::string msg;
		std::getline(std::cin, msg);
		if (msg == "q") {
			quit = true;
		}

		msg += '\n';

		WriteFile(hPipeW,
			msg.c_str(),//"Hello pipe\n",
			msg.length() + 1,
			&dwWritten,
			NULL);
	}
	CloseHandle(hPipeW);
	printf("CPP - Write Stream Closed\n");
}

File: PipeServerTest/pipe_comm.py
import time
import sys
import threading
import win32pipe, win32file, pywintypes

TENSOR_STREAM = "\\\\.\\pipe\\TENSORFLOW_DATA_STREAM"
ADDIN_STREAM = "\\\\.\\pipe\\ADDIN_DATA_STREAM"

quit = False

def Read():
    global quit

    hPipeR = win32file.INVALID_HANDLE_VALUE
    hPipeR = win32pipe.CreateNamedPipe(
        ADDIN_STREAM,
        win32pipe.PIPE_ACCESS_DUPLEX,
        win32pipe.PIPE_TYPE_BYTE | win32pipe.PIPE_READMODE_BYTE | win32pipe.PIPE_WAIT,
        1,
        65536,
        65536,
        0,
        None)

    while hPipeR != win32file.INVALID_HANDLE_VALUE and not quit:
        try:
            print("waiting for client")
            sys.stdout.flush()
            win32pipe.ConnectNamedPipe(hPipeR, None)
            print("got client")
            sys.stdout.flush()

            while True:
                resp = win32file.ReadFile(hPipeR, 1024, None)
                print(f"message: {resp}")
                sys.stdout.flush()


        except pywintypes.error as e:
            if e.args[0] == 2:
                print("no pipe, trying again in a sec")
                sys.stdout.flush()
                time.sleep(1)
            elif e.args[0] == 109:
                print("broken pipe")
                sys.stdout.flush()
                quit = True
        finally:
            win32pipe.DisconnectNamedPipe(hPipeR)

    win32file.CloseHandle(hPipeR)
    print("PY - Read Stream Closed")
    sys.stdout.flush()

def Write():
    global quit

    hPipeW = win32file.INVALID_HANDLE_VALUE

    while not quit:
        try:
            hPipeW = win32file.CreateFile(
                TENSOR_STREAM,
                win32file.GENERIC_READ | win32file.GENERIC_WRITE,
                0,
                None,
                win32file.OPEN_EXISTING,
                0,
                None)

            while not quit:
                strData = input()
                if strData == 'q':
                    quit = True

                data = str.encode(strData+'\n')
                win32file.WriteFile(
				    hPipeW,
				    data
			    )

        except pywintypes.error as e:
            if e.args[0] == 2:
                print("no pipe, trying again in a sec")
                sys.stdout.flush()
                time.sleep(1)
            elif e.args[0] == 109:
                print("broken pipe")
                sys.stdout.flush()
                quit = True

    win32file.CloseHandle(hPipeW)
    print("PY - Write Stream Closed")
    sys.stdout.flush()

try:
    w = threading.Thread(name="Writer", target=Write)
    w.start()
except:
    print("Server Thread Error")
    sys.stdout.flush()

try:
    r = threading.Thread(name="Reader", target=Read)
    r.start()
except:
    print("Client Thread Error")
    sys.stdout.flush()

threads = []
threads.append(w)
threads.append(r)
for t in threads:
    t.join()
